### PR TITLE
4.x - Custom DNS resolvers for Nima

### LINKED ISSUE #4896: Custom DNS resolvers for Nima
Create similar functionality available in Netty to support custom DNS resolvers.

Round-robin impl needs to be provided as well.

### PR DESCRIPTION
**Pull request by Verdent**
Added custom DNS resolving support to Nima client
Added Round Robin implementation

Fixes #4896

### CODE REVIEW

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java [hunk 1]
```diff
@@ -217,7 +218,13 @@ private Http2Headers prepareHeaders(WritableHeaders<?> headers) {
 
     private Http2ClientStream reserveStream() {
         if (explicitConnection == null) {
-            ConnectionKey connectionKey = new ConnectionKey(uri.scheme(), uri.authority(), uri.host(), uri.port(), tls);
+            ConnectionKey connectionKey = new ConnectionKey(uri.scheme(),
+                                                            uri.authority(),
```
Comment (tomas-langer) [inline nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java:222]: `authority` is a combination of `host` and `port`, we should not use both at the same time.

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/DefaultDnsResolverProvider.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+package io.helidon.nima.webclient;
+
+import io.helidon.nima.webclient.spi.DnsResolver;
+import io.helidon.nima.webclient.spi.DnsResolverProvider;
+
+/**
+ * Provider of the {@link DefaultDnsResolver} instance.
+ */
+public class DefaultDnsResolverProvider implements DnsResolverProvider {
```
Comment (tomas-langer) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/DefaultDnsResolverProvider.java:25]: If this is an implementation of service loader, please add explicit default constructor with javadoc describing that it should only be used by service loader.
If we expect users to create instances, add `public static Defa... create()`

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/NoDnsResolverProvider.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+package io.helidon.nima.webclient;
+
+import io.helidon.nima.webclient.spi.DnsResolver;
+import io.helidon.nima.webclient.spi.DnsResolverProvider;
+
+/**
+ * Provider of the {@link NoDnsResolver} instance.
+ */
+public class NoDnsResolverProvider implements DnsResolverProvider {
```
Comment (tomas-langer) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/NoDnsResolverProvider.java:25]: Same as above - if service loader, please add public constructor with javadocs.
Also this should contain a `@Weight` or implement `Weighted`, to have lower weight than other resolvers, so when multiple are on classpath, this one does not get used

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/RoundRobinDnsResolverProvider.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+package io.helidon.nima.webclient;
+
+import io.helidon.nima.webclient.spi.DnsResolver;
+import io.helidon.nima.webclient.spi.DnsResolverProvider;
+
+/**
+ * Provider of the {@link RoundRobinDnsResolver} instance.
+ */
+public final class RoundRobinDnsResolverProvider implements DnsResolverProvider {
```
Comment (tomas-langer) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/RoundRobinDnsResolverProvider.java:25]: As above for other service loader implementations. Also add weights (if we have any preferred order of loading)

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/spi/DnsResolverProvider.java [hunk 5]
```diff
@@ -14,12 +14,25 @@
...
  */
 
-package io.helidon.nima.http2.webclient;
+package io.helidon.nima.webclient.spi;
 
-import io.helidon.nima.common.tls.Tls;
+/**
+ * Provider interface for custom DNS resolvers.
+ */
+public interface DnsResolverProvider {
```
Comment (tomas-langer) [inline nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/spi/DnsResolverProvider.java:22]: This is an SPI and it is called `Provider`, yet I do not see any `uses` statement in `module-info.java`. 
If this is an omission, please fix.
If this is by design (and users are expected to provide an instance), we need to use factory methods for implementations and not default constructors.

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java (hunk 1 above)
Comment (Verdent) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/ClientRequestImpl.java:222]: fixed

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/DefaultDnsResolverProvider.java (hunk 2 above)
Comment (Verdent) [reply nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/DefaultDnsResolverProvider.java:25]: fixed

File: nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/NoDnsResolverProvider.java (hunk 3 above)
Comment (Verdent) [reply nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/NoDnsResolverProvider.java:25]: fixed

**Review by tomas-langer — APPROVED** [review]

**Pull request merged**